tracer/test/Datadog.Trace.ClrProfiler.Managed.Tests/AutoInstrumentation/RabbitMQ/CachedBasicPropertiesHelperTests.cs

```diff
@@ -0,0 +1,66 @@
+﻿// <copyright file="CachedBasicPropertiesHelperTests.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.RabbitMQ;
+using FluentAssertions;
+using RabbitMQ.Client;
+using Xunit;
+
+namespace Datadog.Trace.ClrProfiler.Managed.Tests.AutoInstrumentation.RabbitMQ;
+
+public class CachedBasicPropertiesHelperTests
+{
+    [Fact]
+    public void CreateHeaders_With_Null_BasicProperties_Throws()
+    {
+        // the BasicProperties(IReadOnlyBasicProperties) constructor throws NullReferenceException if parameter is null
+        var func = () => CachedBasicPropertiesHelper<BasicProperties>.CreateHeaders(null!);
+        func.Should().ThrowExactly<NullReferenceException>();
+    }
+
+    [Fact]
+    public void CreateHeaders_BasicProperties_With_Null_Headers()
+    {
+        var originalProperties = new BasicProperties();
+
+        var newProperties = CachedBasicPropertiesHelper<BasicProperties>.CreateHeaders(originalProperties);
+
+        newProperties.Should().NotBeNull();
+        newProperties.Headers.Should().BeNull();
+    }
+
+    [Fact]
+    public void CreateHeaders_BasicProperties_With_Empty_Headers()
+    {
+        var headers = new Dictionary<string, object>();
+
+        var originalProperties = new BasicProperties
+        {
+            Headers = headers
+        };
+
+        var newProperties = CachedBasicPropertiesHelper<BasicProperties>.CreateHeaders(originalProperties);
+
+        newProperties.Should().NotBeNull();
+        newProperties.Headers.Should().BeSameAs(originalProperties.Headers)
+                     .And.Subject.Should().BeEmpty();
+    }
+
+    [Fact]
+    public void CreateHeaders_BasicProperties_With_Headers()
+    {
+        var originalProperties = new BasicProperties();
+        originalProperties.Headers ??= new Dictionary<string, object>();
+        originalProperties.Headers["key1"] = "value1";
+
+        var newProperties = CachedBasicPropertiesHelper<BasicProperties>.CreateHeaders(originalProperties);
+
+        newProperties.Should().NotBeNull();
```

Comment:
For some reason I was thinking we still needed to copy the headers over when we create a new one but I guess not

Reply:
We are using the `BasicProperties(IReadOnlyBasicProperties)` constructor, which copies everything from from the "read only" properties into the mutable one, so we don't have to copy anything over ourselves:

https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/85fd42fc6e245dc68901802967acbccc216b5980/projects/RabbitMQ.Client/BasicProperties.cs#L92C17-L92C18

(In C++ this is called a [copy constructor](https://en.cppreference.com/w/cpp/language/copy_constructor). The term is not common in .NET, even when we use similar patterns, like `BasicProperties` does in this case.)